### PR TITLE
Refactor CI by using wheel packages

### LINKED ISSUE #865: Test and CI refinement
> No files were found with the provided path: /tmp/graphscope/. No artifacts will be uploaded.
- [x] The upload-artifact uploads nothing, please double check the logs location @lidongze0629 
> macOS-latest workflows will use macOS-11 soon. For more details, see https://github.com/actions/virtual-environments/issues/4060
- [ ] Pre-check the compatibility of the dependencies of macOS-11, (low pri, @acezen )
- [x] unify the test scripts for local, use `script/test.sh` as local and minimum test in CI, as we discussed. @lidongze0629 
- [x] clean up the files in `.github/workflows`, remove download-etcd.sh, install-minikube.sh, graphscope.Dockerfile @lidongze0629 
- [x] use `make` command we defined in makefile to build. @lidongze0629 
- [ ] cache the rust built dependencies @siyuan0322 
- [x] `manager` artifact. @lidongze0629 
- [x] Unify `k8s/graphscope.Dcokerfile`and `.github/workflow/graphscope.Dockerfile`

### LINKED ISSUE #918: Improve the CI & docker distribution
**Is your feature request related to a problem? Please describe.**

We will need to keep three image:

1. graphscope-runtime: manylinux + deps
2. graphscope-vineyard: graphscope-runtime + vineyard
3. graphscope: manylinux + java + openmpi + pip installed package

Note that,

+ `graphscope` will be the released image
+ `graphscope-vineyard` will be development environement, will be capable to build store, run store ci, and for daily developement

After that, we have the following CI process:

1. base CI

   1. run `make install` in graphscope-vineyard, resulting a temporary `graphscope-ci:${GIT-SHA}` image

       this step will happen in a self-hosted machine for speed

   2. in the `graphscope-ci:${GIT-SHA}` image, run the following CI:

       1. c++ tests
       2. python tests (with current `NETWORKX=OFF`)
       3. python tests (with current `NETWOPKX=ON`)
       4. gie-test
       5. build wheels, generate the `graphscope` image, then run the k8s ci

    note that the above five steps will be parallelized, and the `make install` process is exactly the same with the steps to setup a local development+testing environment.

2. store-ci

    (I'm not very familar with this part, @siyuan0322 could correct me)

    1. build the store engine and executor in `graphscope-vineyard`, and run CI

3. nightly-ci

    1. run local CI on Ubuntu & Mac every day, to test the compabilities.

**Additional context**

After that we can

1. reduce the complexity of CI infra
2. reduce the time of running CI
3. reduce the size of final released image

### LINKED ISSUE #1056: [BUG] check_gremlin_server_ready can hang up forever
**Describe the bug**

The following code for checking if GIE is ready in coordinator cannot handle timeout correctly, as `client.submit(...)` may hang up for an invalid/failed endpoint.

A possible fix would be put the check loop (includes the `client.submit`) into a threaded executor and return errors when the future/promise timeout.

```python
def check_gremlin_server_ready(endpoint):
    from gremlin_python.driver.client import Client

    if "MY_POD_NAME" in os.environ:
        # inner kubernetes env
        if endpoint == "localhost" or endpoint == "127.0.0.1":
            # now, used in mac os with docker-desktop kubernetes cluster,
            # which external ip is 'localhost' when service type is 'LoadBalancer'
            return True

    client = Client(f"ws://{endpoint}/gremlin", "g")
    error_message = ""
    begin_time = time.time()
    while True:
        try:
            client.submit("g.V().limit(1)").all().result()
        except Exception as e:
            error_message = str(e)
        else:
            client.close()
            return True
        time.sleep(3)
        if time.time() - begin_time > INTERAVTIVE_INSTANCE_TIMEOUT_SECONDS:
            client.close()
            raise TimeoutError(f"Gremlin check query failed: {error_message}")
```

**Additional context**

### PR DESCRIPTION
**Pull request by lidongze0629**
## What do these changes do?

- Reduce the complexity of CI infra
- Reduce the time of running CI
- Reduce the size of the final released image

## Related issue number

Fixes #918 
Fixes #865 
Fixes #1056

### CODE REVIEW

File: .github/workflows/build-graphscope-wheels-macos.yml [hunk 1]
```diff
@@ -50,10 +50,9 @@ jobs:
         path: graphscope.tar.gz
         retention-days: 5
 
-  test:
+  build-client-wheels:
     if: ${{ github.ref == 'refs/heads/main' && github.repository == 'alibaba/GraphScope' }}
     runs-on: ${{ matrix.os }}
```
Comment (acezen) [inline .github/workflows/build-graphscope-wheels-macos.yml:55]: use [macOS-10.15] directly ?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -41,32 +49,44 @@ jobs:
...
           gae-python:
+            - 'proto/**'
             - 'analytical_engine/**'
             - 'python/graphscope/analytical/**'
             - 'python/graphscope/client/**'
             - 'python/graphscope/dataset/**'
             - 'python/graphscope/deploy/**'
             - 'python/graphscope/framework/**'
-            - 'python/graphscope/proto/**'
             - 'python/tests/unittest/**'
```
Comment (acezen) [inline .github/workflows/ci.yml:59]: add a path pattern `coordinator/**`

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -41,32 +49,44 @@ jobs:
...
           networkx:
             - 'analytical_engine/apps/**'
             - 'analytical_engine/frame/**'
             - 'analytical_engine/core/**'
             - 'python/graphscope/nx/**'
+          gie-function-test:
+            - 'interactive_engine/**'
```
Comment (acezen) [inline .github/workflows/ci.yml:66]: add `python/graphscope/interactive/**`

File: .github/workflows/gae.yml [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  push:
+    branches:
+      - main
+    paths:
+      - 'analytical_engine/**'
+  pull_request:
+    branches:
+      - main
+    paths:
+      - 'analytical_engine/**'
```
Comment (acezen) [inline .github/workflows/gae.yml:15]: maybe add a path ignore `**.md` would be better?

File: .github/workflows/gss.yml [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+      - '.github/workflows/gss-yml'
+  pull_request:
+    branches:
+      - main
+    paths:
+      - 'proto/**'
+      - 'interactive_engine/**'
+      - 'python/graphscope/client/**'
+      - 'charts/graphscope-store/**'
+      - '.github/workflows/gss-yml'
```
Comment (acezen) [inline .github/workflows/gss.yml:23]: Ditto, add a path ignore to ignore md file

**Review by siyuan0322 — APPROVED** [review]

File: .github/workflows/build-graphscope-wheels-macos.yml (hunk 1 above)
Comment (lidongze0629) [reply .github/workflows/build-graphscope-wheels-macos.yml:55]:
```
runs-on: macOS-10.15
```

File: .github/workflows/ci.yml (hunk 2 above)
Comment (lidongze0629) [reply .github/workflows/ci.yml:59]: done

File: .github/workflows/ci.yml (hunk 3 above)
Comment (lidongze0629) [reply .github/workflows/ci.yml:66]: done

File: .github/workflows/gae.yml (hunk 4 above)
Comment (lidongze0629) [reply .github/workflows/gae.yml:15]: done

File: .github/workflows/gss.yml (hunk 5 above)
Comment (lidongze0629) [reply .github/workflows/gss.yml:23]: done

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -41,32 +49,44 @@ jobs:
...
 
-  build-gae:
-    runs-on: ubuntu-20.04
+  build-wheels:
+    # Require the user id of the self-hosted is 1001, # which may need to be
+    # configured manually when a new self-hosted runenr is added.
```
[the remark this answers is not in the document]
Comment (lidongze0629) [reply .github/workflows/ci.yml:70]: done

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -120,180 +140,87 @@ jobs:
...
-    - uses: actions/upload-artifact@v2
-      with:
-        name: gie-${{ github.sha }}
-        path: gie.tar
+          client.tar.gz
+          graphscope.tar.gz
         retention-days: 5
 
-  build-gle:
+  minimize-test:
```
[the remark this answers is not in the document]
Comment (lidongze0629) [reply .github/workflows/ci.yml:165]:
```
mini-test:
    xxxx
```

File: .github/workflows/build-graphscope-wheels-linux.yml [hunk 8]
```diff
@@ -53,52 +53,38 @@ jobs:
...
+        sudo make graphscope-image
+        sudo docker tag graphscope/graphscope:${SHORT_SHA} \
+            registry.cn-hongkong.aliyuncs.com/graphscope/graphscope:nightly
+
+    - name: Release Image
+      run: |
+        echo ${{ secrets.ALIYUN_TOKEN }} | sudo docker login --username=grape_dev registry.cn-hongkong.aliyuncs.com --password-stdin
+        sudo docker push registry.cn-hongkong.aliyuncs.com/graphscope/graphscope:nightly
+
+    - name: Clean
```
[the remark this answers is not in the document]
Comment (lidongze0629) [reply .github/workflows/build-graphscope-wheels-linux.yml:86]: done

File: .github/workflows/build-graphscope-wheels-macos.yml [hunk 9]
```diff
@@ -90,23 +85,6 @@ jobs:
...
-        popd
-
-        # run test
-        pip3 install pytest "tensorflow<=2.5.2" --user;
-        export PATH=${HOME}/.local/bin:${PATH}
-        cd ${GITHUB_WORKSPACE}/python
-        python3 -m pytest -s -v ./tests/local/test_standalone.py
-
     - name: Upload Artifact
       if: always()
```
[the remark this answers is not in the document]
Comment (lidongze0629) [reply .github/workflows/build-graphscope-wheels-macos.yml:89]: remove it

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -408,498 +300,238 @@ jobs:
...
+
+          # download dataset
+          git clone -b master --single-branch --depth=1 https://github.com/7br/gstest.git ${GS_TEST_DIR}
+
+      - name: Run Function Test
+        env:
+          GS_TEST_DIR: ${{ github.workspace }}/gstest
+        run: |
+          cd ${GITHUB_WORKSPACE}/interactive_engine/tests
+          ./function_test.sh 8111 1 registry.cn-hongkong.aliyuncs.com/graphscope/graphscope:${SHORT_SHA}
```
[the remark this answers is not in the document]
Comment (lidongze0629) [reply .github/workflows/ci.yml:406]: done

File: k8s/precompile.py [hunk 11]
```diff
@@ -129,7 +129,7 @@ def compile_graph():
```
Comment (acezen) [inline k8s/precompile.py:129]:
```suggestion
    property_frame_template = "vineyard::ArrowFragment<{},{}>"
    projected_frame_template = "gs::ArrowProjectedFragment<{},{},{},{}>"
    flattened_frame_template = "gs::ArrowFlattenedFragment<{},{},{},{}>"
    dynamic_projected_frame_template = "gs::DynamicProjectedFragment<{},{}>"
```

Comment (acezen) [inline k8s/precompile.py:129]:
```suggestion
    flattened_frame_template = "gs::ArrowFlattenedFragment<{},{},{},{}>"
    dynamic_projected_frame_template = "gs::DynamicProjectedFragment<{},{}>"
```

File: k8s/precompile.py [hunk 12]
```diff
@@ -129,7 +129,7 @@ def compile_graph():
 
     oid_types = ["int64_t", "std::string"]
     vid_types = ["uint64_t"]
-    vdata_types = ["grape::EmptyType"]
+    vdata_types = ["int64_t", "grape::EmptyType"]
     edata_types = ["grape::EmptyType", "int64_t", "double"]
     graph_classes = []
```
Comment (acezen) [inline k8s/precompile.py:135]: replace the below code with
```python
    for oid in oid_types:
        for vid in vid_types:
            graph_class = property_frame_template.format(oid, vid)
            graph_classes.append(graph_class)

    for oid in oid_types:
        for vid in vid_types:
            for vdata in vdata_types:
                for edata in edata_types:
                    graph_class = projected_frame_template.format(
                        oid, vid, vdata, edata
                    )
                    flattend_graph_class = flattened_frame_template.format(
                        oid, vid, vdata, edata
                    )
                    graph_classes.append(graph_class)
                    graph_classes.append(flattend_graph_class)

    for vdata in vdata_types:
        for edata in edata_types:
            graph_class = dynamic_projected_frame_template.format(vdata, edata)
            graph_classes.append(graph_class)
```

Comment (acezen) [reply k8s/precompile.py:135]: supplement the compiled app:
```python
def compile_cpp_pie_app():
    template = "gs::ArrowProjectedFragment<{},{},{},{}>"
    flatten_template = "ga::ArrowFlattendFragment<{},{},{},{}"
    dynamic_template = "ga::DynamicProjectedFragment<{},{}"
    luee = template.format(
        "int64_t", "uint64_t", "grape::EmptyType", "grape::EmptyType"
    )
    luel = template.format("int64_t", "uint64_t", "grape::EmptyType", "int64_t")
    lued = template.format("int64_t", "uint64_t", "grape::EmptyType", "double")
    fluee = flatten_template.format(
        "int64_t", "uint64_t", "grape::EmptyType", "grape::EmptyType"
    )
    fluel = flatten_template.format("int64_t", "uint64_t", "grape::EmptyType", "int64_t")
    flued = flatten_template.format("int64_t", "uint64_t", "grape::EmptyType", "double")
    dtee = dynamic_template.format("grape::EmptyType", "grape::EmptyType")
    dtel = dynamic_template.format("grape::EmptyType", "int64_t")
    dted = dynamic_template.format("grape::EmptyType", "double")

    targets = [
        ("pagerank", luee),
        ("pagerank", luel),
        ("wcc", luee),
        ("wcc", luel),
        ("cdlp", luee),
        ("cdlp", luel),
        ("bfs", luee),
        ("bfs", luel),
        ("sssp", luel),
        ("sssp", lued),
        ("kcore", luee),
        ("kcore", dtee),
        ("triangles", luee),
        ("sssp_average_length", luee),
        ("sssp_average_length", luel),
        ("sssp_average_length", lued),
        ("sssp_average_length", dtee),
        ("sssp_average_length", dtel),
        ("sssp_average_length", dted),
        ("pagerak_nx", fluee),
        ("pagerak_nx", dtee),
        ("hits", luee),
        ("hits", fluee),
        ("hits", dtee),
        ("degree_centrality", luee),
        ("degree_centrality", fluee),
        ("degree_centrality", dtee),
        ("eigenvector_centrality", luee),
        ("eigenvector_centrality", luel),
        ("eigenvector_centrality", lued),
        ("eigenvector_centrality", fluee),
        ("eigenvector_centrality", fluel),
        ("eigenvector_centrality", flued),
        ("eigenvector_centrality", dtee),
        ("eigenvector_centrality", dtel),
        ("eigenvector_centrality", dted),
        ("katz_centrality", luee),
        ("katz_centrality", luel),
        ("katz_centrality", lued),
        ("katz_centrality", fluee),
        ("katz_centrality", fluel),
        ("katz_centrality", flued),
        ("katz_centrality", dtee),
        ("katz_centrality", dtel),
        ("katz_centrality", dted),
        ("clustering", luee),
        ("clustering", fluee),
        ("clustering", dtee),
        ("transitivity", luee),
        ("transitivity", fluee),
        ("transitivity", dtee),
        ("avg_clustering", luee),
        ("avg_clustering", fluee),
        ("avg_clustering", dtee),
    ]

    with multiprocessing.Pool() as pool:
        pool.map(cmake_app, targets)
```

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -200,14 +200,14 @@ jobs:
...
-      - name: Run Minimize Test
+      - name: Run Mini Test
         env:
           GS_TEST_DIR: ${{ github.workspace }}/gstest
         run: |
           # download dataset
           git clone -b master --single-branch --depth=1 https://github.com/7br/gstest.git ${GS_TEST_DIR}
 
-          python3 -m pytest -s -v python/tests/minitest
+          python3 -m pytest -s -v $(dirname $(python3 -c "import graphscope; print(graphscope.__file__)"))/tests
```
Comment (siyuan0322) [inline .github/workflows/ci.yml:210]: This grammar is a little bit weird... Why this effort?

Comment (lidongze0629) [reply .github/workflows/ci.yml:210]: Need to test graphscope installed by `pip install`, which exists in `/syspath/site-packages/graphscope`

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -120,180 +149,87 @@ jobs:
...
+        uses: actions/upload-artifact@v2
+        with:
+          name: gie-log
+          path: /tmp/gs/
 
   python-unittest:
     runs-on: ubuntu-20.04
-    needs: [build-gae, build-gle, changes]
+    needs: [build-wheels, changes]
     if: ${{ needs.changes.outputs.gae-python == 'true' }}
```
Comment (acezen) [inline .github/workflows/ci.yml:229]:
```suggestion
    if: ${{ needs.changes.outputs.gae-python == 'true' || github.ref == 'refs/heads/main' }}
```

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -373,15 +282,11 @@ jobs:
 
   networkx-test:
     runs-on: ubuntu-20.04
-    needs: [build-gae, changes]
+    needs: [build-wheels, changes]
     if: ${{ needs.changes.outputs.networkx == 'true' }}
```
Comment (acezen) [inline .github/workflows/ci.yml:286]:
```suggession
    if: ${{ needs.changes.outputs.networkx == 'true'  || github.ref == 'refs/heads/main' }}

File: .github/workflows/ci.yml [hunk 16]
```diff
@@ -408,490 +313,246 @@ jobs:
...
-        pushd /usr/local/dist/ && pip3 install ./*.whl && popd
-        pushd /opt/graphscope/dist/ && pip3 install ./*.whl && popd
-        python3 -m pip install -r python/requirements.txt
-        python3 -m pip install -r python/requirements-dev.txt
-        pushd python && python3 setup.py build_proto && popd
-        pushd coordinator && python3 setup.py build_builtin && popd
-
-    - name: Basic test
+    - name: Basic Test
       if: steps.nx-filter.outputs.basic == 'true'
```
Comment (acezen) [inline .github/workflows/ci.yml:346]:
```suggestion
      if: steps.nx-filter.outputs.basic == 'true' || github.ref == 'refs/heads/main'
```

File: .github/workflows/ci.yml [hunk 17]
```diff
@@ -408,490 +313,246 @@ jobs:
...
-    # so it requires the host is able to run docker without sudo, which may need to be configured manually when a new
-    # self-hosted runner is added.
-    runs-on: self-hosted
-    defaults:
-      run:
-        shell: bash --noprofile --norc -eo pipefail {0}
-    container:
-      image: registry.cn-hongkong.aliyuncs.com/graphscope/graphscope-vineyard:v0.3.9
+    needs: [build-wheels, changes]
+    if: ${{ needs.changes.outputs.gie-function-test == 'true' }}
```
Comment (acezen) [inline .github/workflows/ci.yml:375]:
```suggestion
    if: ${{ needs.changes.outputs.gie-function-test == 'true' || github.ref == 'refs/heads/main' }}
```

File: .github/workflows/ci.yml [hunk 18]
```diff
@@ -343,16 +336,73 @@ jobs:
...
         DEPLOYMENT: ${{ matrix.deployment }}
         GS_TEST_DIR: ${{ github.workspace }}/gstest
       run: |
         python3 -m pytest --exitfirst -s -v \
             $(dirname $(python3 -c "import graphscope; print(graphscope.__file__)"))/nx/tests
 
+  networkx-algo-and-generator-test:
+    runs-on: ubuntu-20.04
+    needs: [build-wheels, changes]
+    if: ${{ needs.changes.outputs.networkx == 'true'  || github.ref == 'refs/heads/main' }}
```
Comment (acezen) [inline .github/workflows/ci.yml:350]: push to main run basic test is enough
```suggestion
    if: ${{ needs.changes.outputs.networkx == 'true' }}
```

File: .github/workflows/ci.yml [hunk 19]
```diff
@@ -343,16 +336,73 @@ jobs:
...
+
+        # install pytest
+        python3 -m pip install pytest
+
+        # download dataset
+        git clone -b master --single-branch --depth=1 https://github.com/7br/gstest.git ${GS_TEST_DIR}
+
     - name: Algorithm test
-      if: steps.nx-filter.outputs.algorithm == 'true'
+      if: steps.nx-filter.outputs.algorithm == 'true' || github.ref == 'refs/heads/main'
```
Comment (acezen) [inline .github/workflows/ci.yml:405]:
```suggestion
      if: steps.nx-filter.outputs.algorithm == 'true'
```

File: .github/workflows/ci.yml [hunk 20]
```diff
@@ -361,7 +411,7 @@ jobs:
             $(dirname $(python3 -c "import graphscope; print(graphscope.__file__)"))/nx/algorithms/tests/builtin
 
     - name: Generator test
-      if: steps.nx-filter.outputs.generator == 'true'
+      if: steps.nx-filter.outputs.generator == 'true' || github.ref == 'refs/heads/main'
```
Comment (acezen) [inline .github/workflows/ci.yml:414]:
```suggestion
      if: steps.nx-filter.outputs.generator == 'true'
```

File: .github/workflows/ci.yml [hunk 21]
```diff
@@ -120,180 +149,87 @@ jobs:
...
+        with:
+          name: gie-log
+          path: /tmp/gs/
 
   python-unittest:
     runs-on: ubuntu-20.04
-    needs: [build-gae, build-gle, changes]
-    if: ${{ needs.changes.outputs.gae-python == 'true' }}
+    needs: [build-wheels, changes]
+    if: ${{ needs.changes.outputs.gae-python == 'true' || github.ref == 'refs/heads/main' }}
```
Comment (acezen) [inline .github/workflows/ci.yml:229]:
```sugesstion
   if:  needs.changes.outputs.gae-python == 'true' || ${{ github.ref == 'refs/heads/main' && github.repository == 'alibaba/GraphScope' }}
```

File: .github/workflows/ci.yml [hunk 22]
```diff
@@ -302,86 +238,55 @@ jobs:
...
         file: ./python/coverage.xml
         fail_ci_if_error: true
 
-  networkx-test:
+  networkx-basic-test:
     runs-on: ubuntu-20.04
-    needs: [build-gae, changes]
-    if: ${{ needs.changes.outputs.networkx == 'true' }}
+    needs: [build-wheels, changes]
+    if: ${{ needs.changes.outputs.networkx == 'true'  || github.ref == 'refs/heads/main' }}
```
Comment (acezen) [inline .github/workflows/ci.yml:286]:
```suggestion
   if:  needs.changes.outputs.networkx == 'true' || ${{ github.ref == 'refs/heads/main' && github.repository == 'alibaba/GraphScope' }}
```

File: .github/workflows/ci.yml [hunk 23]
```diff
@@ -398,500 +303,307 @@ jobs:
...
-        kubectl delete pod graphscope-test-rpc-service --wait=false || true
+        python3 -m pytest --exitfirst -s -v \
+            $(dirname $(python3 -c "import graphscope; print(graphscope.__file__)"))/nx/generators/tests
 
   gie-test:
     runs-on: self-hosted
-    needs: [build-gae, build-gie, build-gle, build-manager]
+    needs: [build-wheels, changes]
+    if: ${{ needs.changes.outputs.gie-function-test == 'true' }}
+    if: ${{ needs.changes.outputs.gie-function-test == 'true' || github.ref == 'refs/heads/main' }}
```
Comment (acezen) [inline .github/workflows/ci.yml:426]:
```suggestion
   if:  needs.changes.outputs.gie-function-test == 'true' || ${{ github.ref == 'refs/heads/main' && github.repository == 'alibaba/GraphScope' }}
```

**Comment by sighingnow** [comment]
@lidongze0629 Could you help to bump up the required vineyard version to v0.3.11?

Thanks~

File: .github/workflows/ci.yml [hunk 24]
```diff
@@ -398,500 +303,307 @@ jobs:
...
-        helm uninstall graphscope || true
-        kubectl delete pod graphscope-test-rpc-service --wait=false || true
+        python3 -m pytest --exitfirst -s -v \
+            $(dirname $(python3 -c "import graphscope; print(graphscope.__file__)"))/nx/generators/tests
 
   gie-test:
     runs-on: self-hosted
-    needs: [build-gae, build-gie, build-gle, build-manager]
+    needs: [build-wheels, changes]
+    if: ${{ needs.changes.outputs.gie-function-test == 'true' }}
```
Comment (acezen) [inline .github/workflows/ci.yml:425]: duplicated.
```sugesstion
```

**Review by sighingnow — APPROVED** [review]
LGTM.

**Pull request merged**